[ENH] Add "multipart DWI" acquisitions and refactor DWI specifications (#624)
bids-standard/bids-specification: src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md

gradient information (`.bval` and `.bvec` sidecar files) to be stored.

Otherwise, if some gradient information is associated to the single-band diffusion
image and a multi-band diffusion image also exists,  the `acq-<label>` key/value pair
MUST be used to distinguish both images.
In such case two files could have the following names:
`sub-01_acq-singleband_dwi.nii.gz` and `sub-01_acq-multiband_dwi.nii.gz`,
however the user is free to choose any other label than `singleband` and
`multiband` as long as they are consistent across subjects and sessions.

### REQUIRED gradient orientation information

The bvec and bval files are in the
[FSL format](https://fsl.fmrib.ox.ac.uk/fsl/fslwiki/FDT/UserGuide#DTIFIT):
The bvec files contain 3 rows with n space-delimited floating-point numbers
(corresponding to the n volumes in the relevant NIfTI file).
The first row contains the *x* elements, the second row contains the *y* elements and
the third row contains the *z* elements of a unit vector in the direction of the applied
diffusion gradient, where the *i*-th elements in each row correspond together to
the *i*-th volume, with `[0,0,0]` for non-diffusion-weighted volumes.

> Inherent to the FSL format for bvec specification is the fact that the coordinate system of

Reviewer: The FSL bvec system is a more complicated than this. It is only true [for some image orientations](https://www.mrtrix.org/2016/04/15/bug-in-fsl-bvecs-handling/). BIDS will need to decide whether they really want to be 100% FSL-compatible or stick to the image axis convention.

Author: Totally agree. My plan is to rebase your #352 on this after this is accepted (I believe this refactor should be less controversial), and then push that PR to the endline.

[gradient information (`.bval` and `.bvec` sidecar files)](#required-gradient-orientation-information)
to be stored.

Otherwise, if some gradient information is associated to the single-band diffusion
image and a multi-band diffusion image also exists, the `acq-<label>` key/value pair
MUST be used to distinguish both images.
In such a case, two files could have the following names:
`sub-01_acq-singleband_dwi.nii.gz` and `sub-01_acq-multiband_dwi.nii.gz`.
The user is free to choose any other label than `singleband` and
`multiband`, as long as they are consistent across subjects and sessions.

### REQUIRED gradient orientation information

The REQUIRED gradient orientation information corresponding to a DWI acquisition
MUST be stored using `.bval` and `.bvec` pairs of files.
The `.bval` and `.bvec` files can be saved on any level of the directory structure
and thus define those values for all sessions and/or subjects in one place (see
Inheritance principle).

As an exception to the [common principles](../02-common-principles.md#definitions)
that parameters are constant across runs, the gradient table information (stored

> within the `.bval` and `.bvec` files) MAY change across DWI runs.

Author: @tsalo please let us know if you think this addresses the issue you pointed out, by making explicit the exception (as I commented before, the exception is currently rather implicit.)

@bids-standard/raw-mri-dwi WDYT?

       sub-<label>[_ses-<label>][_acq-<label>][_dir-<label>][_run-<index>]_sbref.json
```

If more than one run of the same acquisition and direction has been acquired, the
[`run-<index>`](../99-appendices/09-entities.md#run) key/value pair MUST be used:
`_run-1`, `_run-2`, `_run-3` (and so forth.)
When there is only one scan of a given acquisition and direction, the run key MAY be
omitted.
The [`run-<index>`](../99-appendices/09-entities.md#run) key/value pair is RECOMMENDED
to encode the splits of multipart DWI scans (see [below](#multipart-split-dwi-schemes).)

The OPTIONAL [`acq-<label>`](../99-appendices/09-entities.md#acq)
key/value pair corresponds to a custom label the user may use to
distinguish different sets of parameters.

The OPTIONAL [`dir-<label>`](../99-appendices/09-entities.md#dir)
key/value pair corresponds to a custom label the user may use to
distinguish different sets of phase-encoding directions.

**Combining multi- and single-band acquisitions**.
The single-band reference image MAY be stored with suffix `sbref` (for example,

> `dwi/sub-control01_sbref.nii[.gz]`,) as long as the image has no corresponding

Reviewer 2: ```suggestion
`dwi/sub-control01_sbref.nii[.gz]`) as long as the image has no corresponding
```

[gradient information (`.bval` and `.bvec` sidecar files)](#required-gradient-orientation-information)
to be stored.

Otherwise, if some gradient information is associated to the single-band diffusion
image and a multi-band diffusion image also exists, the `acq-<label>` key/value pair

> MUST be used to distinguish both images.

Reviewer 2: I'm assuming so, but just to confirm, does this match current practice? This seems to be a new explicit requirement, although I see that there's never been `_sbref.bval`.

Author: If you see a `_sbref.bval` file then it should be filled with zeros. There could be a use case for nonzero (but very low b-vals << 1000) but why don't use `_dwi` which is more appropriate?

This does match current practice, and as you were pointing, it was implicit in the spec. I just made it explicit, if that makes sense.

Reviewer: There's one large case, which is siemens scanners can't make a b=0 only series without  b > 0 scan included. So the sbref will include a few b=0 images followed by a random b=1000. +1 for optional gradient files for sbrefs

Author: I think the BIDS curator has two options in that case:

- Store a `_dwi` file, although there are not enough orientations to fit a diffusion model this is indeed a *diffusion-weighted* MRI.
- During conversion, drop the b > 0 volume to store a `_sbref` file.

Allowing `_sbref.bval` adds quite a bit of complexity to PyBIDS and software and irregularity (w.r.t. sbrefs under `func/`), while IMHO misrepresenting the nature of the file.

Would you reconsider your +1 to having optional gradient files for sbrefs?

Reviewer: Dropping the b > 0 images would be a processing step, so the sbrefs would no longer be "raw" like the other images. This is actually the same issue that the reverse phase encoding direction images have - the `fmaps/*_epi` images require .bval and .bvec files to be interpreted correctly for distortion correction. 

These are major BIDS limitations that are prevent dmri workflows from using real BIDS and require users put the extra gradient information files into .bidsignore. I think there is another PR open for this, though. I'd be happy if there was even a sidecar field that has the bvals in it, since the gradient directions are irrelevant fir sbrefs and epi fieldmaps.

Author: > Dropping the b > 0 images would be a processing step, so the sbrefs would no longer be "raw" like the other images.

Dropping the b > 0 would be the same as people are actually doing with nonsteady states in functional MRI (at some places), for instance. Or if you intendedly leave out of conversion a whole bunch of acquisitions that are otherwise already in the DICOM structure. I'm not saying anyone should do this, but it is an option - this is outside of BIDS purview.

>  This is actually the same issue that the reverse phase encoding direction images have - the fmaps/*_epi images require .bval and .bvec files to be interpreted correctly for distortion correction.

I don't think this is the same issue. If you have diffusion weights (i.e., nonzero b-values) then you are indeed acquiring DWI - and these files should not be under `fmap/` regardless of whether you can fit a diffusion model or not. I *think* BIDS has sufficient specifications to write these files under `dwi/` as `_dwi.nii` files with corresponding gradient info. 

The problem of estimating a fieldmap using these DWIs is a completely different issue, which is actually addressed in #622.

Finally, accepting `sbref.bval` files would probably not be backward compatible (and a big change I would leave for another PR to keep this one with a focused scope).

> These are major BIDS limitations that are prevent dmri workflows from using real BIDS and require users put the extra gradient information files into .bidsignore.

Can you point at some publicly shared dataset?

In such a case, two files could have the following names:
`sub-01_acq-singleband_dwi.nii.gz` and `sub-01_acq-multiband_dwi.nii.gz`.
The user is free to choose any other label than `singleband` and
`multiband`, as long as they are consistent across subjects and sessions.

### REQUIRED gradient orientation information

The REQUIRED gradient orientation information corresponding to a DWI acquisition

> MUST be stored using `.bval` and `.bvec` pairs of files.

Reviewer 2: I can find no examples of `bvals`/`bvecs` in OpenNeuro, and I do not think the validator has ever recognized those extensions.

> MUST be stored using `.bval` and `.bvec` pairs of files.

Author: Related to this. Just wanted to add that after this is merged, we should push with @mattcieslak for #352. After merge I'll take care of rebasing that one.

> MUST be stored using `.bval` and `.bvec` pairs of files.

Author: (btw, I'm with @effigies here and I don't think we should allow `.[bval|bvec][|s]`, the validator is really useful in indicating the user that those files are not part of the spec)

The most relevant implication for this choice is that any rotations applied to the DWI data
also need to be applied to the corresponding `.bvec` file.

Example of `.bvec` file, with *N*=6, with two *b*=0 volumes in the beginning:

```Text
0 0 0.021828 -0.015425 -0.70918 -0.2465
0 0 0.80242 0.22098 -0.00063106 0.1043
0 0 -0.59636 0.97516 -0.70503 -0.96351
```

The `.bval` file contains the *b*-values (in s/mm<sup>2</sup>) corresponding to the
volumes in the relevant NIfTI file), with 0 designating *b*=0 volumes, space-delimited.

Example of `.bval` file, corresponding to the previous `.bvec` example:

```Text
0 0 2000 2000 1000 1000
```

### Multipart (split) DWI schemes

Some MR schemes cannot be acquired directly by some scanner devices,
requiring to generate several DWI runs that were originally meant to belong
in a single one.
For instance, some GE scanners cannot collect more than &asymp;160 volumes
in a single run under fast-changing gradients, so acquiring *HCP-style*
diffusion images will require splitting the DWI scheme in several runs.
Because researchers will generally optimize the data splits, these will likely

> not be directly concatenable anymore.

Reviewer 3: "concatenable" is a mouthful, maybe:
```suggestion
not be able to be directly concatenated.
```

    sub-1_dir-PA_run-2_dwi.nii.gz  # dwi_1
```

If, conversely, the resarcher wanted to store two multipart scans, one possibility
is to combine matching phase-encoding directions:

```Text
sub-<label>/[ses-<label>/]         # MultipartID
  dwi/
    sub-1_dir-AP_run-1_dwi.nii.gz  # dwi_1
    sub-1_dir-AP_run-2_dwi.nii.gz  # dwi_1
    sub-1_dir-PA_run-1_dwi.nii.gz  # dwi_2
    sub-1_dir-PA_run-2_dwi.nii.gz  # dwi_2
```

Alternatively, the researcher's intent could be combining opposed phase-encoding
runs instead:

```Text
sub-<label>/[ses-<label>/]         # MultipartID
  dwi/
    sub-1_dir-AP_run-1_dwi.nii.gz  # dwi_1
    sub-1_dir-AP_run-2_dwi.nii.gz  # dwi_2
    sub-1_dir-PA_run-1_dwi.nii.gz  # dwi_1
    sub-1_dir-PA_run-2_dwi.nii.gz  # dwi_2
```

The `MultipartID` metadata MAY be used with the
[`acq-<label>`](../99-appendices/09-entities.md#acq) key/value pair, for example:

```Text
sub-<label>/[ses-<label>/]             # MultipartID
  dwi/
    sub-1_acq-shell1_run-1_dwi.nii.gz  # dwi_1
    sub-1_acq-shell1_run-2_dwi.nii.gz  # dwi_2
    sub-1_acq-shell2_run-1_dwi.nii.gz  # dwi_1
    sub-1_acq-shell2_run-2_dwi.nii.gz  # dwi_2
```

### Other RECOMMENDED metadata

> At minimum, the `PhaseEncodingDirection` and `TotalReadoutTime` metadata

Reviewer 2: "Minimum" + "RECOMMENDED" feels like a mismatch.

```suggestion
The `PhaseEncodingDirection` and `TotalReadoutTime` metadata
```

Author: This was meant to say that amongst all the metadata entries we could recommend, these two are really necessary if one wants to be able to perform certain processing (just in case someone wants to recommend something in regards, not just removing the awkward combo).

to encode the splits of multipart DWI scans (see [below](#multipart-split-dwi-schemes).)

The OPTIONAL [`acq-<label>`](../99-appendices/09-entities.md#acq)
key/value pair corresponds to a custom label the user may use to
distinguish different sets of parameters.

The OPTIONAL [`dir-<label>`](../99-appendices/09-entities.md#dir)
key/value pair corresponds to a custom label the user may use to
distinguish different sets of phase-encoding directions.

**Combining multi- and single-band acquisitions**.
The single-band reference image MAY be stored with suffix `sbref` (for example,
`dwi/sub-control01_sbref.nii[.gz]`) as long as the image has no corresponding
[gradient information (`[*_]dwi.bval` and `[*_]dwi.bvec` sidecar files)](#required-gradient-orientation-information)
to be stored.

Otherwise, if some gradient information is associated to the single-band diffusion
image and a multi-band diffusion image also exists, the `acq-<label>` key/value pair
MUST be used to distinguish both images.
In such a case, two files could have the following names:
`sub-01_acq-singleband_dwi.nii.gz` and `sub-01_acq-multiband_dwi.nii.gz`.
The user is free to choose any other label than `singleband` and

> `multiband`, as long as they are consistent across subjects and sessions.

Reviewer: How will the pipelines know that the sbref is intended to be used as the anatomical reference for the multiband sequence? Does this go in the sidecar?

Author: This particular is out of the scope of the refactor - let's take this one to the finish line and then see if extending the `IntendedFor` to sbrefs could do.

The `[*_]dwi.bval` and `[*_]dwi.bvec` files MAY be saved on any level of the directory structure
and thus define those values for all sessions and/or subjects in one place (see
[the inheritance principle](../02-common-principles.md#the-inheritance-principle)).

As an exception to the [common principles](../02-common-principles.md#definitions)
that parameters are constant across runs, the gradient table information (stored
within the `[*_]dwi.bval` and `[*_]dwi.bvec` files) MAY change across DWI runs.

**Gradient orientation file formats**.
The `[*_]dwi.bval` and `[*_]dwi.bvec` files MUST follow the
[FSL format](https://fsl.fmrib.ox.ac.uk/fsl/fslwiki/FDT/UserGuide#DTIFIT):
The `[*_]dwi.bvec` file contains 3 rows with *N* space-delimited floating-point numbers
(corresponding to the *N* volumes in the corresponding NIfTI file.)
The first row contains the *x* elements, the second row contains the *y* elements and
the third row contains the *z* elements of a unit vector in the direction of the applied
diffusion gradient, where the *i*-th elements in each row correspond together to
the *i*-th volume, with `[0,0,0]` for *non-diffusion-weighted* (also called *b*=0 or *low-b*)
volumes.

> Following the FSL format for the `[*_]dwi.bvec` specification, the coordinate system of

Reviewer: The official FSL format for bvecs is more complicated than that and if the bvecs are truly aligned with the image axes many software packages will misinterpret the bvecs. For example an LPS+ image axis will result in flipped y bvecs in FSL and mrtrix, but not DSI Studio or Dipy.

It makes more sense to either say it's "similar to the FSL format, with bvecs aligned to the image axes" or to make it clear that bvecs are *exactly* the FSL format and will intrinsically flip bvecs for some non LAS+ image orientations. The latter seems definitely less BIDSy because it's additional complexity to meet what is arguably a bug in a specific software package rather than to simply describe the data. This problem goes away with the RASB file format in the other PR

Reviewer 3: Actually! Yes, I wanted to flag this.   

I realise this is maybe out of scope of this PR, but note that Appendix 8 mentions 3 types of coordinate systems, "device coordinates" (scanner), "file format coordinates" (DICOM), annd "head" coordinates (for example, NIFTI).

Maybe I can offer a change that links up with the definitions in that appedix?  See next...

Author: Similarly, here I just wanted to:
- Add the `MultipartID` concept
- Refactor the specs of DWI to be more consistent with the functional MRI (a bit more developed than DWI for historical reasons) and making explicit some caveats very implicit in the current spec.

I think #352 provides a great opportunity to refine what is said about the FSL format, make a clear route for deprecation and resolve the problem it originally addressed.

Author: I have accepted the suggestion (adding a link to the coordinate systems is helpful and does not entail any big change to the gradients). That said, I still think we should make the deep revision of gradients in #352.

Reviewer: sounds good. Let me know when you get to #352 and I am happy to help

> the *b* vectors MUST be defined with respect to the participant (in other words, defined
> by the axes of the corresponding `_dwi` NIfTI file) and not the magnet's coordinate system.

Reviewer 3: ```suggestion
the *b* vectors MUST be defined with respect to the head coordinantes in the corresponding `_dwi` NIfTI file and not the scanner's device coordinate system (see [Coordinate systems](08-coordinate-systems.md)).
```

Reviewer 3: And... btw, is "scanner's device coordinate system" satisfactorily clear in `../08-coordinate-systems.md`?  All that is said is
>  all data acquisition starts with "device coordinates" (scanner), which does not have to be identical to the initial "file format coordinates" (DICOM), which are again different from the "head" coordinates (for example, NIFTI).

Author: Thanks for this suggestion. I've taken the liberty of editing your suggestion to fix a little typo ("coordinante") and introduce new lines (I had also fixed the link in parallel to your edit).

Regarding this: as I commented above, I think we should not make very deep changes regarding the gradients, as I think it will more clear for everyone to tackle them in full within #352.

The first row contains the *x* elements, the second row contains the *y* elements and
the third row contains the *z* elements of a unit vector in the direction of the applied
diffusion gradient, where the *i*-th elements in each row correspond together to
the *i*-th volume, with `[0,0,0]` for *non-diffusion-weighted* (also called *b*=0 or *low-b*)
volumes.
Following the FSL format for the `[*_]dwi.bvec` specification, the coordinate system of
the *b* vectors MUST be defined with respect to the coordinate system defined by
the header of the corresponding `_dwi` NIfTI file and not the scanner's device
coordinate system (see [Coordinate systems](../99-appendices/08-coordinate-systems.md)).
The most relevant implication for this choice is that any rotations applied to the DWI data
also need to be applied to the *b* vectors in the `[*_]dwi.bvec` file.

Example of `[*_]dwi.bvec` file, with *N*=6, with two *b*=0 volumes in the beginning:

```Text
0 0 0.021828 -0.015425 -0.70918 -0.2465
0 0 0.80242 0.22098 -0.00063106 0.1043
0 0 -0.59636 0.97516 -0.70503 -0.96351
```

The `[*_]dwi.bval` file contains the *b*-values (in s/mm<sup>2</sup>) corresponding to the

> volumes in the relevant NIfTI file), with 0 designating *b*=0 volumes, space-delimited.

Reviewer 4: depending on the scanner and sequence some of the BVALs might have very small values in locations where 0 is expected (Siemens sometimes adds 5, 10, 15 as an example). Similarly, values for non-zero BVALs can be not-rounded off say 990 or 995 instead of 1,000 or 2020, 2010 or 1995 instead of 2000. I think this is fine for 80% of the situations but it might require changes in the future.

Author: I think @mattcieslak initiated the right place for this conversation with #352. For a cleaner diff that assumes this PR has been already merged, please use https://github.com/oesteban/bids-specification/compare/enh/dwi-combinations...mattcieslak:dwi_tsv

and thus define those values for all sessions and/or subjects in one place (see
Inheritance principle).

## Multipart (split) DWI schemes

Some MR schemes cannot be acquired directly by some scanner devices,
requiring to generate several DWI runs that were originally meant to belong

> in a single one.

Reviewer 5: I'm not sure if we would want to do this, but would it make sense to use `split` in this case instead of `run`? `run` implies that the parameters are the same, so based on the current definition under **Common principles**, I think you'd want to use `acq` to distinguish the two files (e.g., `acq-first160dirs`, `acq-second160dirs`).

Author: At the moment, two `_dwi` runs with all other BIDS entities being equal can differ in there bvecs/bvals specifications. Therefore, you can perfectly have `run-1` and `run-2` encoding the first 160 and the second 160 orientations respectively.

I agree that whether bvecs/bvals should be constant across runs is rather unclear at the moment (and the validator does not check this particular for bvecs/bvals) - I can send a separate PR to clarify this. One thing is sure - there's a lot of data out there with several `run-` indices each with a different gradient table (and I don't think it would be wise more explicitly disallowing this).

That said, the idea for this PR is not to limit the utilization of the proposed new key (`MultipartID`) across runs, also across different acquisitions and/or directions. If the proposal is not clear as regards this, please let me know.

Author: @tsalo I have added https://github.com/bids-standard/bids-specification/pull/624/files#diff-d668661006216379ab2a5c5ff7061b8aR450-R452 to refer to this issue more explicitly in the gradient-table section

Author: Please let me know further concerns after the latest commits which sideways have contributed to clarify this.

Author: I tagged you above, where the exception to the common principles is made explicit. The rest of the discussion I think can be followed with the bulk of the conversation (esp. Chris' comments). Marking resolved - please mark unresolved if you disagree.

As an exception to the [common principle](../02-common-principles.md#definitions)
that parameters are constant across runs, the gradient table information (stored
within the `.bval` and `.bvec` files) MAY change across DWI runs.

### Multipart (split) DWI schemes

Some MR schemes cannot be acquired directly by some scanner devices,
requiring to generate several DWI runs that were originally meant to belong
in a single one.
For instance, some GE scanners cannot collect more than &asymp;160 volumes
in a single run under fast-changing gradients, so acquiring *HCP-style*
diffusion images will require splitting the DWI scheme in several runs.

| **Key name**    | **Requirement level** | **Data type** | **Description**                                                                                                                                                                                                      |
| --------------- | --------------------- | ------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | MultipartID     | REQUIRED              | [string][]    | Unique identifier (participant's level) to prescribe DWI segments belonging in a single dataset. Segments can be designated by any combination of `dir-<label>`, `acq-<label>` and/or `run-<index>` key/value pairs. |

Author: will rewrite, thanks for pointing out.

> | MultipartID     | REQUIRED              | [string][]    | Unique identifier (participant's level) to prescribe DWI segments belonging in a single dataset. Segments can be designated by any combination of `dir-<label>`, `acq-<label>` and/or `run-<index>` key/value pairs. |

Author: please check if now it is clearer

JSON example:

```JSON
{
  "PhaseEncodingDirection": "j-",
  "TotalReadoutTime": 0.095,

>   "MultipartID": "dwi_protocol1"

Reviewer: great idea

Reviewer 4: This seems good to me to.

}
```

For instance, the above JSON sidecar defined for a set of files via the
`sub-<label>_dwi.json` file would indicate that all the runs 1 through 4
belong in the same unique image:

```Text
sub-<label>/[ses-<label>/]
    dwi/
       sub-<label>_run-1_dwi.nii.gz
       sub-<label>_run-1_dwi.bval
       sub-<label>_run-1_dwi.bvec
       sub-<label>_run-2_dwi.nii.gz
       sub-<label>_run-2_dwi.bval
       sub-<label>_run-2_dwi.bvec
       sub-<label>_run-3_dwi.nii.gz
       sub-<label>_run-3_dwi.bval
       sub-<label>_run-3_dwi.bvec
       sub-<label>_run-4_dwi.nii.gz
       sub-<label>_run-4_dwi.bval
       sub-<label>_run-4_dwi.bvec

>        sub-<label>_dwi.json

Reviewer 6: Would a json file for each of the individual runs no longer exist? Or would `sub-<label>_dwi.json` contain the unique tags that would trickle down and supersede the tags in the individual run jsons?

Author: This is just to make the example shorter, but the rationale of this new tag `MultipartID` still works with multiple jsons.
